include/swift/SIL/SILRemarkStreamer.h

```diff
@@ -0,0 +1,111 @@
+//===--- SILRemarkStreamer.h - Interface for streaming remarks --*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+/// \file
+/// This file defines the interface used to stream SIL optimization diagnostics
+/// through LLVM's RemarkStreamer interface.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_SIL_SILREMARKSTREAMER_H
+#define SWIFT_SIL_SILREMARKSTREAMER_H
+
+#include "swift/Demangling/Demangler.h"
+#include "swift/SIL/OptimizationRemark.h"
+#include "llvm/Remarks/RemarkStreamer.h"
+
+namespace swift {
+
+class SILRemarkStreamer {
+  llvm::remarks::RemarkStreamer &RS;
+  // Source manager for resolving source locations.
+  const SourceManager &SM;
+  // Temporary objects to avoid allocating memory for every remark when building
+  // strings.
+  mutable std::string TmpRemarkName;
+  mutable std::string TmpFunctionName;
+  /// Convert diagnostics into remark objects.
+  /// The lifetime of the members of the result is bound to the lifetime of
+  /// the SIL remarks.
+  template <typename RemarkT>
+  llvm::remarks::Remark toRemark(const OptRemark::Remark<RemarkT> &R) const;
+
+public:
+  SILRemarkStreamer(llvm::remarks::RemarkStreamer &RS, const SourceManager &SM)
+      : RS(RS), SM(SM) {}
+  /// Emit a remark through the streamer.
+  template <typename RemarkT> void emit(const OptRemark::Remark<RemarkT> &R);
+};
+
+std::pair<std::unique_ptr<llvm::raw_fd_ostream>,
+          std::unique_ptr<SILRemarkStreamer>>
+createSILRemarkStreamer(SILModule &SM, StringRef Filename, StringRef Passes,
+                        StringRef Format, DiagnosticEngine &DE,
+                        SourceManager &SourceMgr);
+
+// Implementation for template member functions.
+
+// OptRemark type -> llvm::remarks::Type
+template <typename RemarkT> static llvm::remarks::Type toRemarkType() {
+  if (std::is_same<RemarkT, OptRemark::RemarkPassed>::value)
```

Comment:
Please do not write code like this! It breaks LLVM style. Use early returns!

LLVM Coding Standard: https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return